conduit/data/datasets/base.py

```diff
@@ -62,21 +92,28 @@ def logger(self) -> logging.Logger:
             self._logger = init_logger(self.__class__.__name__)
         return self._logger
 
+    @overload
+    def _sample_x(self, index: IndexType, *, coerce_to_tensor: Literal[True] = ...) -> Tensor:
+        ...
+
+    @overload
+    def _sample_x(self, index: IndexType, *, coerce_to_tensor: Literal[False] = ...) -> LoadedData:
+        ...
+
     def _sample_x(
-        self, index: Union[int, List[int], slice], *, coerce_to_tensor: bool = False
-    ) -> Tensor:
-        self.x
+        self, index: IndexType, *, coerce_to_tensor: bool = False
+    ) -> Union[LoadedData, Tensor]:
```

Comment:
Why is the return type not
```suggestion
    ) -> X:
```